### PR TITLE
feat(server): Add basePath config to host from a directory

### LINKED ISSUE #697: Add support for custom root paths
Howdy,

Short version: 

I'm trying to host my Fresh app at `https://my-domain-name/homepage`. How do I tell Fresh to use the `/homepage` routing extension , analogous to putting the "homepage" address into package.json for a React app?

Longer version:

I have an EC2 set up running Apache that has some Plotly Dash apps, and I'm trying to use Fresh to build a simple home page for it. (I have very little web-dev experience, this is more a learning exercise).  I want this page to be accessible at `https://my-domain-name/homepage`.

In my Apache config I've added the usual reverse proxy statements:

`ProxyPass  /homepage  http://127.0.0.1:8000`
`ProxyPassReverse /homepage  http://127.0.0.1:8000`

(I've got this set up for some Plotly Dash apps and they all work fiine)

I created a new Fresh boilerplate repo (with the Lemon image and the Counter) and ran it with `deno task start` (so running on port 8000 by default).

When I look at `localhost:8000` on this EC2, all is well. App is running normally.

When I look at `http://my-domain-name/homepage`, the html objects show up but the islands and static assets don't load. When I inspct the network traffic, the first request "GET my-domain-name/homepage" returns 200. But all the subsequent requests to fetch the static assets and JS files are like:

"GET my-domain-name/_frsh/refresh.js"
"GET my-domain-name/logo.svg?__frsh_c=927807f6-b7ac-4bb4-b2ab-d55181719490" 
etc

returning 404s because they're all without the /homepage/ extension.  So how do I tell Fresh to put the `/homepage/` extension on these follow up requests?

Cheers.

### PR DESCRIPTION
**Pull request by deer**
closes #697 
supersedes #1294

This was quite a lot to redo, since everything has changed. Eventually I just gave up and started over. So somehow I ended up with an implementation where the value in the config is what the user types in (e.g. `basePath: "/foo/bar"`) whereas your implementation has a line like `basePath = opts.router?.basePath.slice(1) + "/";`.

### CODE REVIEW

File: docs/canary/concepts/server-configuration.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+This setting allows you to serve a Fresh app from sub-path of a domain. A value
+of `/foo/bar` would serve the app from `http://localhost:8000/foo/bar` instead
+of `http://localhost:8000/` for example.
+
+The `basePath` will be automatically applied to absolute links in your app. For
+example, when the `basePath` is `/foo/bar`, linking to `/about` will
+automatically become `/foo/bar/about`.
+
+```js
```
Comment (marvinhagemeister) [inline docs/canary/concepts/server-configuration.md]: Super tiny nit:
```suggestion
```jsx
```

**Comment by deer** [comment]
It also seems like a more holistic approach to windows issues is needed. I felt like I was smashing them where I found them, rather than having a good overall approach. Perhaps this is something for a followup PR, since I don't have anything concrete to say.

**Review by marvinhagemeister — APPROVED** [review]
Nice, thanks for bringing this over the finish line 🙌

**Comment by marvinhagemeister** [comment]
Yeah it's fine for now for as long as the windows tests pass.

**Comment by deer** [comment]
```
./tests/www/navigate_test.ts => www - docs navigation ... FAILED (19s)

 ERRORS 

www - docs navigation => ./tests/www/navigate_test.ts:3:6
error: TimeoutError: waiting for function failed: timeout 2000ms exceeded
      const timeoutError = new TimeoutError(
                           ^
    at new WaitTask (https://deno.land/x/puppeteer@16.2.0/vendor/puppeteer-core/puppeteer/common/IsolatedWorld.js:796:28)
    at IsolatedWorld.waitForFunction (https://deno.land/x/puppeteer@16.2.0/vendor/puppeteer-core/puppeteer/common/IsolatedWorld.js:662:22)
    at Frame.waitForFunction (https://deno.land/x/puppeteer@16.2.0/vendor/puppeteer-core/puppeteer/common/Frame.js:530:36)
    at Page.waitForFunction (https://deno.land/x/puppeteer@16.2.0/vendor/puppeteer-core/puppeteer/common/Page.js:2975:29)
    at waitForText (file:///Users/runner/work/fresh/fresh/tests/test_utils.ts:410:16)
    at eventLoopTick (ext:core/01_core.js:178:11)
    at async file:///Users/runner/work/fresh/fresh/tests/www/navigate_test.ts:12:5
    at async withPageName (file:///Users/runner/work/fresh/fresh/tests/test_utils.ts:232:7)
    at async file:///Users/runner/work/fresh/fresh/tests/www/navigate_test.ts:5:3

 FAILURES 

www - docs navigation => ./tests/www/navigate_test.ts:3:6

FAILED | 305 passed (113 steps) | 1 failed | 2 ignored (5m12s)
```

**Pull request merged**